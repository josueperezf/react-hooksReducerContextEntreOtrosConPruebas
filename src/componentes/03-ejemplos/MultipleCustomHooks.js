import React from 'react'
import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch';
import '../02-useEffect/effect.css';
export const MultipleCustomHooks = () => {

    const {contador, incrementar} = useCounter(1);
    // console.log(contador);
    const url = `https://www.breakingbadapi.com/api/quotes/${contador}`;
    const { loading, data } = useFetch(url);
    /**
     * con un ! en el data, si la misma viene null, el resultado seria true,
     * con dos !! en el data, si la data viene null, el resultado es falso,
     * con dos !! si el resultado es true, mejor dicho, si tiene data, entonces '&&' obtenga de data[0]
     */
    // si existe la data extrae la posicion cero
    // console.log(data);
    
    const {author, quote} = !!data && data[0];
    return (
        <div>
            <h1>Frases de breaking bad</h1>
            <hr/>
            {
                (loading) ?
                    <div className="alert alert-info text-center"> Loading... </div>
                :
                    (
                        <blockquote className="blockquote text-end">
                            <p className='mb-0' >{quote}</p>
                            <footer className="blockquote-footer">{author}</footer>
                        </blockquote>
                    )
            }
            <button
                className="btn btn-primary"
                onClick={()=>incrementar(1) }
                > Siguiente frase</button>
        </div>
    )
}
