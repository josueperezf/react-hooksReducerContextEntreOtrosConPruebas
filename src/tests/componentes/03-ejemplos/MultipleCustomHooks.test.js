
// import { renderHook, act } from '@testing-library/react-hooks';
import { shallow } from "enzyme";
import { MultipleCustomHooks } from "../../../componentes/03-ejemplos/MultipleCustomHooks";
import { useFetch } from "../../../hooks/useFetch";
import { useCounter } from "../../../hooks/useCounter";

// no me importa la implementacin de useFetch, solo los resultados, es para emular que ya hizo la llamada http
// con la siguiente llamada evita la llamada del hook usefech
jest.mock('../../../hooks/useFetch');

// la siguiente es para emular tambien la llamada del hook que me sirve de contador
jest.mock('../../../hooks/useCounter');

describe('Pruebas en <MultipleCustomHooks/>', ()=>{

    beforeEach( () => {
        useCounter.mockReturnValue({
            contador:10,
            incrementar:()=>{},
            decrementar:()=>{},
            reset:()=>{},
        });
    });


    test('debe mostrarse correctamente <MultipleCustomHooks/>', () => {
        // suponemos que usefech hizo la llamada al servidor y lo siguiente fue la respuesta
        useFetch.mockReturnValue({
            data:   null,
            loading:true,
            error:  null
        });

        const wrapper = shallow(<MultipleCustomHooks />);
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de mostrar la informacion', () => {

        useFetch.mockReturnValue({
            data:   [{
                author:'Josue',
                quote:'Hola Mundo'
            }],
            loading:false,
            error:  null
        });
        const wrapper = shallow(<MultipleCustomHooks />);
        // console.log(wrapper.html());
        // es para verificar si el alert que se pone cuando la data esta cargando, realmente desaparecio
        expect(wrapper.find('.alert').exists()).toBe(false );
        expect(wrapper.find('.mb-0').text().trim()).toBe('Hola Mundo');
        expect(wrapper.find('footer').text().trim()).toBe('Josue');
    });
    
    
});