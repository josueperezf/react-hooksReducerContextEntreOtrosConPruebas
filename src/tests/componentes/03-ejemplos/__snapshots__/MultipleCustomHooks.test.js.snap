// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks/> debe mostrarse correctamente <MultipleCustomHooks/> 1`] = `
<div>
  <h1>
    Frases de breaking bad
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
     Loading... 
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
     Siguiente frase
  </button>
</div>
`;
